File: python/fan-out/fan-out/handler.py
import os

import requests


TARGET_FUNCTION = "url-check"
SUBMIT_TIMEOUT = 30


def handle(event, context):
    body = (
        event.body.decode()
        if isinstance(event.body, bytes)
        else str(event.body)
    )
    records = [
        record.strip()
        for record in body.strip().splitlines()
        if record.strip()
    ]
    if not records:
        return error(400, "expected one record per line in the request body")

    gateway = os.getenv("gateway_url", "http://gateway.openfaas:8080")
    callback = event.headers.get("X-Callback-Url", "").strip()
    if not callback:
        callback = os.getenv("callback_url", "").strip()

    call_ids = []
    for index, record in enumerate(records):
        try:
            call_id = submit(gateway, record, callback)
        except requests.RequestException as err:
            return error(502, f"record {index + 1} of {len(records)}: {err}")
        except RuntimeError as err:
            return error(502, f"record {index + 1} of {len(records)}: {err}")

        if call_id:
            call_ids.append(call_id)

    response = {
        "submitted": len(records),
        "function": TARGET_FUNCTION,
        "callback": bool(callback),
    }
    if call_ids:
        response["call_ids"] = call_ids

    return {"statusCode": 200, "body": response}


def submit(gateway, record, callback):
    headers = {"Content-Type": "text/plain"}
    if callback:
        headers["X-Callback-Url"] = callback

    response = requests.post(
        f"{gateway.rstrip('/')}/async-function/{TARGET_FUNCTION}",
        data=record.encode(),
        headers=headers,
        timeout=SUBMIT_TIMEOUT,
    )
    if response.status_code != 202:
        raise RuntimeError(
            f"unexpected status {response.status_code} "
            f"from {TARGET_FUNCTION}: {response.text}"
        )

    return response.headers.get("X-Call-Id", "")


def error(status_code, message):
    return {"statusCode": status_code, "body": message}

File: go/fan-out/fan-out/handler.go
package function

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"strings"
	"time"
)

const (
	targetFunction = "url-check"
	submitTimeout  = 30 * time.Second
)

type Response struct {
	Submitted int      `json:"submitted"`
	Function  string   `json:"function"`
	Callback  bool     `json:"callback"`
	CallIDs   []string `json:"call_ids,omitempty"`
}

// Handle takes a HTTP request body and splits it into one record per line.
// Each record is submitted as an asynchronous invocation of the target
// function, then a summary is returned to the caller without waiting for
// the function invocations to complete.
func Handle(w http.ResponseWriter, r *http.Request) {
	input, err := io.ReadAll(r.Body)
	if err != nil {
		http.Error(w, "unable to read request body", http.StatusBadRequest)
		return
	}
	defer r.Body.Close()

	gateway := os.Getenv("gateway_url")
	if gateway == "" {
		gateway = "http://gateway.openfaas:8080"
	}

	// Forward the callback URL to every asynchronous invocation. A header on
	// the batch request overrides the environment variable.
	callback := strings.TrimSpace(r.Header.Get("X-Callback-Url"))
	if callback == "" {
		callback = strings.TrimSpace(os.Getenv("callback_url"))
	}

	records := recordsFromInput(string(input))
	if len(records) == 0 {
		http.Error(
			w,
			"expected one record per line in the request body",
			http.StatusBadRequest,
		)
		return
	}

	submitted := 0
	var callIDs []string

	for i, record := range records {
		callID, err := submit(
			r.Context(), gateway, targetFunction, record, callback,
		)
		if err != nil {
			message := fmt.Sprintf(
				"record %d of %d: %s",
				i+1,
				len(records),
				err,
			)
			http.Error(w, message, http.StatusBadGateway)
			return
		}

		submitted++
		if callID != "" {
			callIDs = append(callIDs, callID)
		}
	}

	res := Response{
		Submitted: submitted,
		Function:  targetFunction,
		Callback:  callback != "",
		CallIDs:   callIDs,
	}

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(res)
}

func recordsFromInput(input string) []string {
	var records []string

	for _, record := range strings.Split(strings.TrimSpace(input), "\n") {
		record = strings.TrimSpace(record)
		if record != "" {
			records = append(records, record)
		}
	}

	return records
}

func submit(
	ctx context.Context,
	gateway string,
	targetFunction string,
	record string,
	callback string,
) (string, error) {
	ctx, cancel := context.WithTimeout(ctx, submitTimeout)
	defer cancel()

	url := strings.TrimRight(gateway, "/") + "/async-function/" + targetFunction
	req, err := http.NewRequestWithContext(
		ctx,
		http.MethodPost,
		url,
		bytes.NewReader([]byte(record)),
	)
	if err != nil {
		return "", fmt.Errorf("unable to invoke %s: %w", targetFunction, err)
	}
	req.Header.Set("Content-Type", "text/plain")
	if callback != "" {
		req.Header.Set("X-Callback-Url", callback)
	}

	res, err := http.DefaultClient.Do(req)
	if err != nil {
		return "", fmt.Errorf("error invoking %s: %w", targetFunction, err)
	}
	defer res.Body.Close()

	if res.StatusCode != http.StatusAccepted {
		out, err := io.ReadAll(res.Body)
		if err != nil {
			return "", fmt.Errorf(
				"unexpected status %d from %s",
				res.StatusCode,
				targetFunction,
			)
		}

		return "", fmt.Errorf(
			"unexpected status %d from %s: %s",
			res.StatusCode,
			targetFunction,
			string(out),
		)
	}

	// the X-Call-Id header can be used to track or cancel the record
	return res.Header.Get("X-Call-Id"), nil
}
